### PR TITLE
server: hardens the remote shard registration

### PR DESCRIPTION
**Pull request by p0lyn0mial**
## Summary
changes the way we register the remote shard so that it handles the following cases:

- the resource has already been registered and is up to date
- the resource has been registered but requires an update
- the resource hasn't been added

all the above cases tolerate any availability issues while connecting to the root shard.

## Related issue(s)

Fixes #

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -244,10 +244,53 @@ func (s *Server) Run(ctx context.Context) error {
 					VirtualWorkspaceURL: s.Options.Extra.ShardVirtualWorkspaceURL,
 				},
 			}
-			if _, err := s.RootShardKcpClusterClient.Cluster(tenancyv1alpha1.RootCluster).TenancyV1alpha1().ClusterWorkspaceShards().Create(goContext(ctx), shard, metav1.CreateOptions{}); err != nil {
-				return err
+
+			if err := wait.PollInfiniteWithContext(goContext(ctx), time.Second, func(ctx context.Context) (bool, error) {
```
Comment (stevekuznetsov) [inline pkg/server/server.go:248]: Should we use an exponential backoff here?

**Review by stevekuznetsov — COMMENTED** [review]
Is there any easy way to test this logic?

File: pkg/server/server.go [hunk 2]
```diff
@@ -244,10 +244,53 @@ func (s *Server) Run(ctx context.Context) error {
...
+					_, err := s.RootShardKcpClusterClient.Cluster(tenancyv1alpha1.RootCluster).TenancyV1alpha1().ClusterWorkspaceShards().Update(ctx, existingShard, metav1.UpdateOptions{})
+					if err != nil {
+						klog.Error("failed updating %q ClusterWorkspaceShard in the root shard, err %v", shard.Name, err)
+						return false, nil
+					}
+					klog.Infof("Updated %q ClusterWorkspaceShard resource in the root shard", shard.Name)
+					return true, nil
+				}
+				klog.Infof("ClusterWorkspaceShard %q resource is up to date in the root shard", shard.Name)
+				return true, nil
```
Comment (sttts) [inline pkg/server/server.go:288]: all this worth a method?

File: pkg/server/server.go [hunk 3]
```diff
@@ -244,10 +244,53 @@ func (s *Server) Run(ctx context.Context) error {
 					VirtualWorkspaceURL: s.Options.Extra.ShardVirtualWorkspaceURL,
 				},
 			}
-			if _, err := s.RootShardKcpClusterClient.Cluster(tenancyv1alpha1.RootCluster).TenancyV1alpha1().ClusterWorkspaceShards().Create(goContext(ctx), shard, metav1.CreateOptions{}); err != nil {
-				return err
+
+			if err := wait.PollInfiniteWithContext(goContext(ctx), time.Second, func(ctx context.Context) (bool, error) {
+				klog.Infof("Getting %q ClusterWorkspaceShard from the root shard", shard.Name)
```
Comment (sttts) [inline pkg/server/server.go:249]:
```suggestion
				klog.Infof("Getting %q ClusterWorkspaceShard from the root workspace", shard.Name)
```

File: pkg/server/server.go [hunk 4]
```diff
@@ -244,10 +244,53 @@ func (s *Server) Run(ctx context.Context) error {
 					VirtualWorkspaceURL: s.Options.Extra.ShardVirtualWorkspaceURL,
```
Comment (sttts) [inline pkg/server/server.go:244]: all these are defaulted already?

File: pkg/server/server.go [hunk 5]
```diff
@@ -244,10 +244,53 @@ func (s *Server) Run(ctx context.Context) error {
...
+				}
+				if len(shard.Spec.ExternalURL) > 0 && existingShard.Spec.ExternalURL != shard.Spec.ExternalURL {
+					existingShard.Spec.ExternalURL = shard.Spec.ExternalURL
+					specChanged = true
+				}
+				if len(shard.Spec.VirtualWorkspaceURL) > 0 && existingShard.Spec.VirtualWorkspaceURL != shard.Spec.VirtualWorkspaceURL {
+					existingShard.Spec.VirtualWorkspaceURL = shard.Spec.VirtualWorkspaceURL
+					specChanged = true
+				}
+				if specChanged {
```
Comment (sttts) [inline pkg/server/server.go:278]: what if we skipped all this checking and just update spec unconditionally? Would be good enough, wouldn't it?

File: pkg/server/server.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/server/server.go:248]: I don't think so. We do it on a server startup, so we want it to be quick. Backing off might delay server initialization.

**Comment by p0lyn0mial** [comment]
> Is there any easy way to test this logic?

yes, you could `./bin/sharded-test-server --number-of-shards=2` on your local machine

File: pkg/server/server.go (hunk 5 above)
Comment (p0lyn0mial) [reply pkg/server/server.go:278]: ah, I have just realized that the fields would be updated by the admission plugin. So yeah it seems to be safe.

File: pkg/server/server.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/server/server.go:288]: I'll move it to a separate hook in the next step. So I'd leave it as it is.

**Comment by stevekuznetsov** [comment]
https://github.com/kcp-dev/kcp/pull/1700#issuecomment-1209062432

I meant a programmatic test ... ;)

**Pull request merged**